ocaml/typing/typeopt.ml

```diff
@@ -801,6 +808,9 @@ let lazy_val_requires_forward env loc ty =
     Misc.fatal_error "Unboxed value encountered inside lazy expression"
   | Float -> Config.flat_float_array
   | Addr | Int -> false
+  | Product cs ->
+    let kind = Jkind.Sort.Const.Product cs in
+    raise (Error (loc, Unsupported_product_in_lazy kind))
```

Comment:
Is this error case tested? I don't remember seeing it.

Reply:
This error should be unreachable if the typechecker is correct (note it gives the scary "Contact us" message).